File: python/leetcode/303.py
class NumArray(object):
    def __init__(self, nums):
        n = len(nums)
        self.dp = [0] * (n + 1)
        for i in range(1, n+1):
            self.dp[i] = nums[i-1] + self.dp[i-1]

    def sumRange(self, i, j):
        return self.dp[j + 1] - self.dp[i]

File: cpp/leetcode/303.cpp
class NumArray {
public:
    NumArray(vector<int> &nums) {
        dp.resize(nums.size() + 1);
        dp[0] = 0;
        for (int i = 1; i <= nums.size(); i++) {
            dp[i] = nums[i - 1] + dp[i - 1];
        }
    }

    int sumRange(int i, int j) {
        return dp[j + 1] - dp[i];
    }

private:
    vector<int> dp;
};